patches/api/0474-WIP-DataKey-API.patch

```diff
@@ -0,0 +1,433 @@
++    }
++
++}
+diff --git a/src/main/java/io/papermc/paper/datakey/PatchedDataKeyMap.java b/src/main/java/io/papermc/paper/datakey/PatchedDataKeyMap.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..f5c8c3c4f6989119934e47f82a120e486108c699
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/datakey/PatchedDataKeyMap.java
+@@ -0,0 +1,31 @@
++package io.papermc.paper.datakey;
++
++import io.papermc.paper.datakey.patch.DataKeyMapPatch;
++import org.jetbrains.annotations.NotNull;
++import javax.annotation.Nullable;
++import java.util.Objects;
++import java.util.Optional;
++
++public interface PatchedDataKeyMap extends DataKeyMap {
++
++    static PatchedDataKeyMap empty() {
++        return DataKeyBridge.Holder.BRIDGE.of(DataKeyBridge.Holder.BRIDGE.empty());
++    }
++
++    static PatchedDataKeyMap of(DataKeyMap map) {
++        return DataKeyBridge.Holder.BRIDGE.of(map);
++    }
++
++    @Nullable
++    <T> T set(DataKey<T> type, @Nullable T value);
++
++    @Nullable
++    <T> T remove(DataKey<T> type);
++
++    @NotNull
++    DataKeyMapPatch asPatch();
++
++    void applyPatch(@NotNull DataKeyMapPatch patch);
++
++    PatchedDataKeyMap clone();
++}
+diff --git a/src/main/java/io/papermc/paper/datakey/patch/DataKeyMapPatch.java b/src/main/java/io/papermc/paper/datakey/patch/DataKeyMapPatch.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..325a597b8d3e0c574b7bc80b7c9d0deeb3b28037
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/datakey/patch/DataKeyMapPatch.java
+@@ -0,0 +1,5 @@
++package io.papermc.paper.datakey.patch;
++
++// TODO: Do we want this?
++public interface DataKeyMapPatch {
++}
+diff --git a/src/main/java/io/papermc/paper/datakey/types/ComponentTypesBridge.java b/src/main/java/io/papermc/paper/datakey/types/ComponentTypesBridge.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..04d908dc4133f853772813b87c4954eae00cfb5d
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/datakey/types/ComponentTypesBridge.java
+@@ -0,0 +1,13 @@
++package io.papermc.paper.datakey.types;
++
++import net.kyori.adventure.util.Services;
++
++public interface ComponentTypesBridge {
++
++    Unbreakable unbreakable(boolean showInTooltip);
++
++    public static class Holder {
++        public static ComponentTypesBridge BRIDGE = Services.service(ComponentTypesBridge.class)
++            .orElseThrow();
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/datakey/types/Unbreakable.java b/src/main/java/io/papermc/paper/datakey/types/Unbreakable.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..d6dc5fa334679767096da9f80740bbc9d9630617
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/datakey/types/Unbreakable.java
+@@ -0,0 +1,15 @@
++package io.papermc.paper.datakey.types;
++
++import org.jetbrains.annotations.Contract;
++
++public interface Unbreakable {
++
++    static Unbreakable of(boolean showInTooltip) {
```

Comment:
this looks ugly

Reply:
Do you prefer something like that?

```suggestion
+    static Unbreakable.Builder builder() {
```

and usage would be smth like that:

```java
Unbreakable.builder().showInTooltip()....build();
```

Reply:
`showInTooltip` should accept a `boolean`.